workers/unity/Packages/com.improbable.gdk.core/Dispatcher.cs

```diff
@@ -0,0 +1,494 @@
+﻿using System;
+using System.Collections.Generic;
+using Improbable.WorkerCore;
+
+namespace Improbable.Gdk.Core
+{
+    internal class DescendingComparer<T> : IComparer<T> where T : IComparable<T>
+    {
+        public int Compare(T x, T y)
+        {
+            return y.CompareTo(x);
+        }
+    }
+
+    // We need to avoid changing the underlying map while iterating over it. For example, if a
+    // callback removes itself, a call to InvokeAll iterates over the map to invoke each callback.
+    // This then indirectly invokes Remove, which makes the iteration crash.
+    //
+    // A simple solution to this problem would be to copy the map for each iteration, but this
+    // allocates a significant amount of memory each time. Instead, we fix this issue by postponing
+    // changes to the map until the end of the top-level call to this object. In the mentioned
+    // example, when InvokeAll indirectly calls Remove, the removal is not done immediately, but
+    // postponed to the end of the (top-level) InvokeAll call.
+    //
+    // The current callbacks are those in the map, plus toAdd, minus toRemove. The changes are
+    // merged via the UpdateGuard.
+    internal class Callbacks<T>
+    {
+        private SortedDictionary<ulong, Action<T>> map = new SortedDictionary<ulong, Action<T>>();
+
+        private SortedDictionary<ulong, Action<T>> mapReversed =
+            new SortedDictionary<ulong, Action<T>>(new DescendingComparer<ulong>());
+
+        private SortedDictionary<ulong, Action<T>> toAdd = new SortedDictionary<ulong, Action<T>>();
```

Comment:
These should be list of tuple I think.

Reply:
Generally it might be worth us making all of these lists. As the most common task is to iterate it's unlikely a sorted dictionary is the best structure.

Reply:
Sorted dictionary will allocate on foreach.

Reply:
Doesn't have to be done here. Will be adding callback stuff in one of my PRs so can go through this then.